### PR TITLE
refactor(gatsby): change bootstrap to use services

### PR DESCRIPTION
**Pull request by ascorbic**
This splits the `bootstrap` command into smaller parts, which it imports from services. A lot of this is in the `initialize` service, which is mostly [a direct TypeScript port](https://github.com/gatsbyjs/gatsby/pull/24816/files#diff-c02ff0d485cf139e3d2e1aba409a5c83) from the relevant section of the old bootstrap file. 

The services that this PR includes are: 

- initialize
- customize-schema
- source-nodes
- build-schema
- create-pages
- create-pages-statefully
- extract-queries
- write-out-requires
- write-out-redirects
- post-bootstrap

### CODE REVIEW

File: packages/gatsby/src/bootstrap/index.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  sourceNodes,
+  buildSchema,
+  writeOutRequires,
+  createPages,
+  createPagesStatefully,
+  extractQueries,
+} from "../services"
+import { Runner } from "./create-graphql-runner"
+import { writeOutRedirects } from "../services/write-out-redirects"
+import { postBootstrap } from "../services/post-bootstrap"
```
Comment (pieh) [inline packages/gatsby/src/bootstrap/index.ts:15]: any reasons why those are imported from their files and not from `"../services"` like every other service?

File: packages/gatsby/src/bootstrap/index.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import { rebuildWithSitePage } from "../schema"
```
Comment (pieh) [inline packages/gatsby/src/bootstrap/index.ts:16]: This one feels werid to actually not be service given that everything else is (other than types and one "startListener").

What qualifies something to be a service really?

File: packages/gatsby/src/bootstrap/index.ts (hunk 1 above)
Comment (ascorbic) [reply packages/gatsby/src/bootstrap/index.ts:15]: TypeScript/VS Code being weird! I'll fix

Comment (pieh) [reply packages/gatsby/src/bootstrap/index.ts:15]: that's minor/NIT thing, so no need to do this quickly, it just gave me a pause ("what's different about those?")

File: packages/gatsby/src/bootstrap/index.ts (hunk 2 above)
Comment (ascorbic) [reply packages/gatsby/src/bootstrap/index.ts:16]: In short: it being used in state machine. I'm not 100% sure whether I'll need to use this one there, but you're right I should probably make it one anyway.

File: packages/gatsby/src/bootstrap/index.ts (hunk 1 above)
Comment (ascorbic) [reply packages/gatsby/src/bootstrap/index.ts:15]: No, it is a mistake. I mostly let it auto-import, and sometimes it does it from the wrong place

File: packages/gatsby/src/services/post-bootstrap.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  reporter.log(``)
+  reporter.info(`bootstrap finished - ${process.uptime().toFixed(3)}s`)
+  reporter.log(``)
+  emitter.emit(`BOOTSTRAP_FINISHED`, {})
+  require(`../redux/actions`).boundActionCreators.setProgramStatus(
+    `BOOTSTRAP_FINISHED`
+  )
+
+  // eslint-disable-next-line no-unused-expressions
+  parentSpan?.finish()
```
[the remark this answers is not in the document]
Comment (ascorbic) [reply packages/gatsby/src/services/post-bootstrap.ts:25]: Good point

File: packages/gatsby/src/services/create-pages-statefully.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
+import { IBuildContext } from "./"
+
+import reporter from "gatsby-cli/lib/reporter"
+import apiRunnerNode from "../utils/api-runner-node"
+
+export async function createPagesStatefully({
+  parentSpan,
+  bootstrapGraphQLFunction,
```
[the remark this answers is not in the document]
Comment (ascorbic) [reply packages/gatsby/src/services/create-pages-statefully.ts:8]: What would you call it? I'm not a big fan of the name, but I found it super confusing having two things called GraphQL Runner, one of which is an object and another which is the actual `graphql()` function, which is why I renamed it.

File: packages/gatsby/src/commands/types.ts [hunk 5]
```diff
@@ -22,6 +24,7 @@ export interface IProgram {
   sitePackageJson: PackageJson
   ssl?: ICert
   graphqlTracing?: boolean
+  setStore?: (store: Store<IGatsbyState, AnyAction>) => void
```
[the remark this answers is not in the document]
Comment (ascorbic) [reply packages/gatsby/src/commands/types.ts:27]: Yes

File: packages/gatsby/src/bootstrap/index.ts (hunk 2 above)
Comment (pieh) [reply packages/gatsby/src/bootstrap/index.ts:16]: I don't think this one will be used in any other place (schema-hot-reloader uses `rebuild` from `src/schema/index`), but I do like "entry points" for each ... "service" to be in same place (`src/services/`) vs some in services and some elsewhere.

Also another NIT - this change as is lost the `update schema` activity wrapping the execution of `rebuildWithSitePage`

```
activity = reporter.activityTimer(`update schema`, {	
    parentSpan: bootstrapSpan,	
  })
```

So if this is moved to service - that would be natural place to re-add it

File: packages/gatsby/src/services/extract-queries.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  const activity = reporter.activityTimer(`onPreExtractQueries`, {
+    parentSpan,
+  })
+  activity.start()
+  await apiRunnerNode(`onPreExtractQueries`, {
+    parentSpan: activity.span,
+  })
+  activity.end()
+
+  await extractQueriesAndWatch({ parentSpan })
```
[the remark this answers is not in the document]
Comment (pieh) [reply packages/gatsby/src/services/extract-queries.ts:19]: Please, not in this PR - all this PR does is aliasing `extractQueries` to `extractQueriesAndWatch`, but it doesn't change implementation.

In insides of that function however - we setup watcher only if ``NODE_ENV === `development` `` (which is heurisitc we use in some places for `gatsby develop` only logic). So it already acts bit different for build and develop (tho setup is convoluted and details/conditions are placed at weird abstraction levels)

[the remark this answers is not in the document]
Comment (wardpeet) [reply packages/gatsby/src/services/extract-queries.ts:19]: Read `later stage` 😄

File: packages/gatsby/src/commands/develop-process.ts [hunk 7]
```diff
@@ -419,10 +419,10 @@ module.exports = async (program: IProgram): Promise<void> => {
   }
 
   // Start bootstrap process.
-  const { graphqlRunner } = await bootstrap(program)
+  const { bootstrapGraphQLFunction } = await bootstrap({ program })
```
[the remark this answers is not in the document]
Comment (pieh) [reply packages/gatsby/src/commands/develop-process.ts:422]: graphqlRunner has some state on itself - creating new one will cause additional (unneeded) work being done (related to materialization etc). Creating one and reusing it is good and we should do it, I think just the naming is suspect here - it's not really `bootstrapGraphQLFunction` - it is `gatsbyNodeGraphqlFunction` (as it's passed to `onPostBuild` etc which definitely is not in bootstrap). Maybe the way it's being passed around is also confusing - maybe there should be singleton and instead of passing it around we reach for it whenever we need it?

File: packages/gatsby/src/services/create-pages-statefully.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (pieh) [reply packages/gatsby/src/services/create-pages-statefully.ts:8]: Related - https://github.com/gatsbyjs/gatsby/pull/24816/files#r440903969

File: packages/gatsby/src/commands/develop-process.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (pieh) [reply packages/gatsby/src/commands/develop-process.ts:422]: > Maybe the way it's being passed around is also confusing - maybe there should be singleton and instead of passing it around we reach for it whenever we need it?

Tho I wouldn't do any of that in this PR (or maybe at all even)

File: packages/gatsby/index.d.ts [hunk 8]
```diff
@@ -1283,29 +1283,6 @@ export type ActivityArgs = {
...
-  log: LogMessageType
-  activityTimer(name: string, activityArgs?: ActivityArgs): ActivityTracker
-  createProgress(
-    text: string,
-    total?: number,
-    start?: number,
-    activityArgs?: ActivityArgs
-  ): ProgressActivityTracker
-}
-
```
Comment (pieh) [inline packages/gatsby/index.d.ts:1308]: Are those changes needed for this PR? those reporter type changes almost feel like should be separate PR?

File: packages/gatsby/src/commands/develop-process.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (ascorbic) [reply packages/gatsby/src/commands/develop-process.ts:422]: Currently it's for inc builds purposes

[the remark this answers is not in the document]
Comment (ascorbic) [reply packages/gatsby/src/commands/develop-process.ts:422]: I definitely prefer `gatsbyNodeGraphqlFunction`

[the remark this answers is not in the document]
Comment (ascorbic) [reply packages/gatsby/src/commands/develop-process.ts:422]: Or maybe `apiRunnerGraphqlFunction`. Any other bikesheds?

File: packages/gatsby/index.d.ts (hunk 8 above)
Comment (ascorbic) [reply packages/gatsby/index.d.ts:1308]: Done: https://github.com/gatsbyjs/gatsby/pull/25045

File: packages/gatsby/src/services/initialize.ts [hunk 9]
```diff
@@ -1,53 +1,59 @@
...
+import { IGatsbyState } from "../redux/types"
+import { IBuildContext } from "./types"
+
+const tracer = globalTracer()
 
 // Show stack trace on unhandled promises.
-process.on(`unhandledRejection`, (reason, p) => {
-  reporter.panic(reason)
+process.on(`unhandledRejection`, reason => {
+  reporter.panic({ id: ``, context: reason } as ErrorMeta)
```
[the remark this answers is not in the document]
Comment (ascorbic) [reply packages/gatsby/src/services/initialize.ts:31]: Changed back, with better typing

File: packages/gatsby/index.d.ts (hunk 8 above)
Comment (sidharthachatterjee) [reply packages/gatsby/index.d.ts:1308]: Merged

**Comment by wardpeet** [comment]
I'll run this locally but looks good to me. Good job!

File: packages/gatsby/src/commands/build.ts [hunk 10]
```diff
@@ -70,8 +70,8 @@ module.exports = async function build(program: IBuildArgs): Promise<void> {
   const buildSpan = buildActivity.span
   buildSpan.setTag(`directory`, program.directory)
 
-  const { graphqlRunner: bootstrapGraphQLRunner } = await bootstrap({
-    ...program,
+  const { gatsbyNodeGraphQLFunction } = await bootstrap({
```
Comment (pvdz) [inline packages/gatsby/src/commands/build.ts:73]: @ascorbic btw when I mean reducing irrelevant changes out, this is the kind. This file has 5 changed lines. Three of them concern a var rename. I bet we can find many of these that are not really relevant to the core of this PR.

You can probably separate out the bootstrap api changes as well. (Export default to export binding, program no longer being spread, returned keys, etc).

File: packages/gatsby/src/services/build-schema.ts [hunk 11]
```diff
@@ -0,0 +1,19 @@
+import { IBuildContext } from "./"
+
+import { build } from "../schema"
+import reporter from "gatsby-cli/lib/reporter"
+
+export async function buildSchema({
+  store,
+  parentSpan,
+}: Partial<IBuildContext>): Promise<void> {
+  if (!store) {
```
Comment (pvdz) [inline packages/gatsby/src/services/build-schema.ts:10]: Wouldn't TS Catch this error? I'm fine leaving the error in, just wondering.

File: packages/gatsby/src/services/create-pages.ts [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+}: Partial<IBuildContext>): Promise<void> {
+  if (!store) {
+    reporter.panic(`store not initialized`)
+  }
+  const activity = reporter.activityTimer(`createPages`, {
+    parentSpan,
+  })
+  activity.start()
+  // const timestamp = Date.now()
+  // const currentPages = new Map<string, IGatsbyPage>(store.getState().pages)
```
Comment (pvdz) [inline packages/gatsby/src/services/create-pages.ts:19]: Did you leave these intentionally? Same for comments further down

File: packages/gatsby/src/services/initialize.ts [hunk 13]
```diff
@@ -1,54 +1,52 @@
-/* @flow */
-
-const _ = require(`lodash`)
```
Comment (pvdz) [inline packages/gatsby/src/services/initialize.ts:3]: I wish github was more contextual and recognize and match these imports :( So much small low hanging fruit there.

File: packages/gatsby/src/services/initialize.ts [hunk 14]
```diff
@@ -1,54 +1,52 @@
...
-import { removeStaleJobs } from "./remove-stale-jobs"
+import { IGatsbyState } from "../redux/types"
+import { IBuildContext } from "./types"
 
 // Show stack trace on unhandled promises.
-process.on(`unhandledRejection`, (reason, p) => {
-  reporter.panic(reason)
+process.on(`unhandledRejection`, (reason: unknown) => {
+  // https://github.com/DefinitelyTyped/DefinitelyTyped/issues/33636
+  reporter.panic((reason as Error) || `Unhandled rejection`)
```
Comment (pvdz) [inline packages/gatsby/src/services/initialize.ts:31]: Probably `new Error('Unhandled rejection')` or something? You could add that there was no explicit message :shrug:

File: packages/gatsby/src/services/initialize.ts [hunk 15]
```diff
@@ -1,54 +1,52 @@
...
+// require(`../bootstrap/log-line-function`)
+
+export async function initialize(
+  context: IBuildContext
+): Promise<{
+  store: Store<IGatsbyState, AnyAction>
+  workerPool: JestWorker
+}> {
+  const args = context.program
+  const { parentSpan } = context
```
Comment (pvdz) [inline packages/gatsby/src/services/initialize.ts:46]: Nit:
```suggestion
  { program: args, parentSpan }: IBuildContext
): Promise<{
  store: Store<IGatsbyState, AnyAction>
  workerPool: JestWorker
}> {
```

File: packages/gatsby/src/services/initialize.ts [hunk 16]
```diff
@@ -1,54 +1,52 @@
...
-
-type BootstrapArgs = {
-  directory: string,
-  prefixPaths?: boolean,
-  parentSpan: Object,
-  graphqlTracing: boolean,
-}
-
-module.exports = async (args: BootstrapArgs) => {
-  const spanArgs = args.parentSpan ? { childOf: args.parentSpan } : {}
```
Comment (pvdz) [inline packages/gatsby/src/services/initialize.ts:50]: I presume this `childOf` stuff is not relevant anymore?

File: packages/gatsby/src/services/initialize.ts [hunk 17]
```diff
@@ -286,19 +282,15 @@ module.exports = async (args: BootstrapArgs) => {
...
-    parentSpan: bootstrapSpan,
+    parentSpan,
   })
   activity.start()
   const srcDir = `${__dirname}/../../cache-dir`
   const siteDir = cacheDirectory
   const tryRequire = `${__dirname}/../utils/test-require-error.js`
   try {
-    await fs.copy(srcDir, siteDir, {
-      clobber: true,
```
Comment (pvdz) [inline packages/gatsby/src/services/initialize.ts:297]: Curious: why was this no longer required? Is that a change that should be it's own PR?

File: packages/gatsby/src/services/initialize.ts [hunk 18]
```diff
@@ -336,25 +328,30 @@ module.exports = async (args: BootstrapArgs) => {
     return undefined
   }
 
-  const ssrPlugins = _.filter(
-    flattenedPlugins.map(plugin => {
+  interface IPluginResolution {
+    resolve: string
+    options: IPluginInfoOptions
+  }
```
Comment (pvdz) [inline packages/gatsby/src/services/initialize.ts:334]: I think interfaces/type definitions should be at the top of a file. Should we add a lint for this?

File: packages/gatsby/src/services/initialize.ts [hunk 19]
```diff
@@ -336,25 +328,30 @@ module.exports = async (args: BootstrapArgs) => {
...
+
+  const isResolved = (plugin): plugin is IPluginResolution => !!plugin.resolve
```
Comment (pvdz) [inline packages/gatsby/src/services/initialize.ts:336]: TIL about `is`. Seems valid. Not quite convinced that's a pattern we should use.

File: packages/gatsby/src/services/post-bootstrap.ts [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+  const activity = reporter.activityTimer(`onPostBootstrap`, {
+    parentSpan,
+  })
+  activity.start()
+  await apiRunnerNode(`onPostBootstrap`, { parentSpan: activity.span })
+  activity.end()
+
+  reporter.log(``)
+  reporter.info(`bootstrap finished - ${process.uptime().toFixed(3)}s`)
+  reporter.log(``)
```
Comment (pvdz) [inline packages/gatsby/src/services/post-bootstrap.ts:18]: I guess this wasn't your change but can't this pre/append newlines to the middle one instead?

File: packages/gatsby/src/services/source-nodes.ts [hunk 21]
```diff
@@ -0,0 +1,61 @@
+import { IBuildContext } from "./"
+import sourceNodesAndRemoveStaleNodes from "../utils/source-nodes"
+import reporter from "gatsby-cli/lib/reporter"
+// import { findChangedPages } from "../utils/check-for-changed-pages"
+// import { IGatsbyPage } from "../redux/types"
```
Comment (pvdz) [inline packages/gatsby/src/services/source-nodes.ts:5]: Comments in this file, probably to delete?

File: packages/gatsby/src/services/build-schema.ts (hunk 11 above)
Comment (ascorbic) [reply packages/gatsby/src/services/build-schema.ts:10]: The context has the store as nullable, because it's not there when the state machine starts. We add it early on, but because the context is always the same type, we need to add an explicit check. Having the runtime check also helps to debug state machine issues

File: packages/gatsby/src/services/create-pages.ts (hunk 12 above)
Comment (ascorbic) [reply packages/gatsby/src/services/create-pages.ts:19]: Yes, because these are temporarily removed just while we're handling this outside the state machine. The main state machine branch has that code it, and I want to make it obvious where it's meant to go back in.

File: packages/gatsby/src/services/initialize.ts (hunk 14 above)
Comment (ascorbic) [reply packages/gatsby/src/services/initialize.ts:31]: `reporter.panic` accepts all manner or arguments. I don't know whether a string or an error is better here tbh.

File: packages/gatsby/src/services/initialize.ts (hunk 17 above)
Comment (ascorbic) [reply packages/gatsby/src/services/initialize.ts:297]: This was an error surfaced when porting to TypeScript. `clobber` is an invalid option (and is the default behaviour anyway). It's ignored, so doesn't cause any harm, but does cause a TS error.

File: packages/gatsby/src/services/initialize.ts (hunk 19 above)
Comment (ascorbic) [reply packages/gatsby/src/services/initialize.ts:334]: It's super useful, because it narrows the type when filtering an array so you don't need to do extra typechecking.

File: packages/gatsby/src/services/post-bootstrap.ts (hunk 20 above)
Comment (ascorbic) [reply packages/gatsby/src/services/post-bootstrap.ts:18]: Yeah, this was how it was before. Following the principle of fewest irrelevant changes, I'd not refactored anything like this.

File: packages/gatsby/src/services/initialize.ts (hunk 17 above)
Comment (wardpeet) [reply packages/gatsby/src/services/initialize.ts:297]: yeah it's now called override but it's true by default

**Comment by ascorbic** [comment]
@pvdz Thanks for the detailed review! Regarding that activity stuff, yes it is annoying to have all that boilerplate, though I'm not sure how much shorter it would be if I abstracted it: I'd only really be able to remove the explicit `start()` call, as everything else would still need passing in, and it would still need stopping. Any ideas for a cleaner way?

**Comment by wardpeet** [comment]
@pdvz Let's not refactor activities. I like the verbosity of it + we might need to think about another way in the future for this. In state machine land these could be side effects

File: packages/gatsby/src/services/initialize.ts (hunk 18 above)
Comment (ascorbic) [reply packages/gatsby/src/services/initialize.ts]: Done

File: packages/gatsby/src/services/post-bootstrap.ts (hunk 20 above)
Comment (ascorbic) [reply packages/gatsby/src/services/post-bootstrap.ts]: Done

File: packages/gatsby/src/commands/types.ts [hunk 22]
```diff
@@ -15,14 +16,15 @@ export interface IProgram {
   port: number
   proxyPort: number
   host: string
-  report: typeof reporter
+  report: Reporter
```
Comment (muescha) [inline packages/gatsby/src/commands/types.ts:19]: somehow this reverts this PR:

- #24941 `refactor(gatsby): convert program reducer to typescript`

```diff
-  report: Reporter
+  report: typeof reporter
```

was this intended?

//cc @tgallacher
